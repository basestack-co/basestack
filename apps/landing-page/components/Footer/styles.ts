import styled from "styled-components";
import Link from "next/link";
import { rem } from "polished";
import { gradientBorderStyles } from "../styles";

export const Container = styled.footer`
  display: flex;
  flex-direction: column;
  margin-top: auto;
  padding: 0 ${({ theme }) => theme.spacing.s5};
  position: relative;
  ${gradientBorderStyles("top")};
`;

export const ContentWrapper = styled.footer`
  display: flex;
  flex-direction: column;
  padding-bottom: ${({ theme }) => theme.spacing.s8};
  max-width: ${rem("1400px")};
  width: 100%;
  margin: 0 auto;
`;

export const NewsletterContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: ${({ theme }) => theme.spacing.s8} 0;
  gap: ${({ theme }) => theme.spacing.s5};

  @media screen and ${({ theme }) => theme.device.max.md} {
    flex-direction: column;
    align-items: flex-start;
    justify-content: flex-start;
  }
`;

export const NewsletterContent = styled.div`
  display: flex;
  flex-direction: column;
`;

export const InputContainer = styled.div`
  display: flex;
  flex-grow: 1;
  justify-content: flex-end;

  @media screen and ${({ theme }) => theme.device.max.md} {
    width: 100%;
    justify-content: flex-start;
  }
`;

export const LinksContainer = styled.div`
  display: grid;
  grid-template-columns: repeat(4, minmax(0, auto));
  width: 100%;
  grid-gap: ${({ theme }) => theme.spacing.s8};
  padding: ${({ theme }) => theme.spacing.s8} 0;
  justify-content: space-between;

  @media screen and ${({ theme }) => theme.device.max.md} {
    grid-template-columns: repeat(2, minmax(0, 1fr));
  }

  @media screen and ${({ theme }) => theme.device.max.sm} {
    grid-template-columns: 1fr;
  }
`;

export const BottomContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding-top: ${rem("25px")};
  gap: ${({ theme }) => theme.spacing.s8};
`;

export const CompanyContainer = styled.div`
  display: flex;
  align-items: center;
`;

export const ListContainer = styled.div`
  display: flex;
  flex-direction: column;
`;

export const List = styled.ul`
  display: flex;
  flex-direction: column;
  margin-top: ${({ theme }) => theme.spacing.s3};
`;

export const ListItem = styled.li`
  display: flex;
  flex-direction: column;

  &:not(:last-child) {
    margin-bottom: ${({ theme }) => theme.spacing.s2};
  }
`;

export const StyledLink = styled(Link)`
  text-decoration: none;
  color: ${({ theme }) => theme.text.color};
  line-height: 1.4;
  font-weight: 400;

  &:hover {
    text-decoration: underline;
  }
`;

export const LogoContainer = styled.div`
  display: inline-flex;
  flex-shrink: 0;
`;
