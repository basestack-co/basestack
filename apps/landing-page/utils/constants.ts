export const sendinblueApiUrl = "https://api.sendinblue.com/v3";
export const redirectionUrl = "https://basestack-landing-page.onrender.com/";
