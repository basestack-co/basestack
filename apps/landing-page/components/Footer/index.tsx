import React from "react";
import { useStore } from "store";
// Router
import { useRouter } from "next/navigation";
// Utils
import { useDarkModeToggle } from "@basestack/hooks";
import { events } from "@basestack/utils";
// Form
import { useForm, Controller, SubmitHandler } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
// Theme
import { useTheme } from "styled-components";
// Components
import {
  Text,
  Button,
  ButtonSize,
  HorizontalRule,
  Logo,
  IconButton,
  Input,
  ButtonVariant,
} from "@basestack/design-system";
import {
  Container,
  InputContainer,
  List,
  ListItem,
  ListContainer,
  RightContainer,
  LeftContainer,
  ContentWrapper,
  MainContent,
  BottomContainer,
  StyledLink,
} from "./styles";
import { z } from "zod";

const products = [
  {
    text: "Feature Flags",
    href: "/flags",
  },
  {
    text: "Forms",
    href: "/forms",
  },
];

const links = [
  {
    text: "Privacy Policy",
    href: "/legal/privacy",
  },
  {
    text: "Cookies Policy",
    href: "/legal/cookies",
  },
  {
    text: "Term of Use",
    href: "/legal/terms",
  },
];

export const FormSchema = z.object({
  email: z
    .string()
    .min(1, "Required field")
    .email("This is not a valid email."),
});

export type FormInputs = z.TypeOf<typeof FormSchema>;

const Footer = () => {
  const theme = useTheme();
  const router = useRouter();
  const { toggleDarkMode } = useDarkModeToggle();

  const setIsDarkMode = useStore((state) => state.setDarkMode);
  const isDarkMode = useStore((state) => state.isDarkMode);

  const {
    control,
    handleSubmit,
    formState: { errors, isSubmitting },
    reset,
  } = useForm<FormInputs>({
    resolver: zodResolver(FormSchema),
    mode: "onChange",
  });

  const onSubmit: SubmitHandler<FormInputs> = async (input: FormInputs) => {
    if (!process.env.NEXT_PUBLIC_FORM_UPDATES_ENDPOINT) return;

    try {
      const res = await fetch(process.env.NEXT_PUBLIC_FORM_UPDATES_ENDPOINT, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ email: input.email }),
      });

      await res.json();

      events.landing.newsletter(
        "Subscribe Success",
        "Subscribe with success to the Newsletter",
      );
      reset();
    } catch (error) {
      const { message } = error as Error;

      events.landing.newsletter("Subscribe Error", message);
    }
  };

  return (
    <Container>
      <HorizontalRule isDarker={!isDarkMode} />
      <ContentWrapper>
        <MainContent>
          <LeftContainer>
            <Logo product="company" isOnDark={isDarkMode} />
            <Text
              size="medium"
              fontWeight={400}
              mt={theme.spacing.s6}
              mb={theme.spacing.s2}
            >
              Get updates directly to your Inbox
            </Text>
            <InputContainer>
              <Controller
                name="email"
                control={control}
                defaultValue=""
                render={({ field }) => (
                  <Input
                    isDarker={!isDarkMode}
                    placeholder="Enter your email"
                    type="email"
                    name={field.name}
                    value={field.value}
                    onChange={field.onChange}
                    onBlur={field.onBlur}
                    isDisabled={isSubmitting}
                    mr={theme.spacing.s2}
                  />
                )}
              />
              <Button
                onClick={handleSubmit(onSubmit)}
                isLoading={isSubmitting}
                size={ButtonSize.Medium}
                variant={ButtonVariant.Tertiary}
                height="44px"
              >
                Subscribe
              </Button>
            </InputContainer>
          </LeftContainer>
          <RightContainer>
            <ListContainer>
              <Text size="medium">Products</Text>
              <List>
                {products.map((link, index) => (
                  <ListItem
                    key={index.toString()}
                    onClick={() => {
                      events.landing.link(link.text, link.href);
                      router.push(link.href);
                    }}
                  >
                    <StyledLink href={link.href}>{link.text}</StyledLink>
                  </ListItem>
                ))}
              </List>
            </ListContainer>

            <ListContainer>
              <Text size="medium">Docs</Text>
              <List>
                {links.map((link, index) => (
                  <ListItem
                    key={index.toString()}
                    onClick={() => {
                      events.landing.link(link.text, link.href);
                      router.push(link.href);
                    }}
                  >
                    <StyledLink href={link.href}>{link.text}</StyledLink>
                  </ListItem>
                ))}
              </List>
            </ListContainer>

            <ListContainer>
              <Text size="medium">Company</Text>
              <List>
                {links.map((link, index) => (
                  <ListItem
                    key={index.toString()}
                    onClick={() => {
                      events.landing.link(link.text, link.href);
                      router.push(link.href);
                    }}
                  >
                    <StyledLink href={link.href}>{link.text}</StyledLink>
                  </ListItem>
                ))}
              </List>
            </ListContainer>
          </RightContainer>
        </MainContent>

        <HorizontalRule isDarker={!isDarkMode} mt={theme.spacing.s8} />
        <BottomContainer>
          <Text size="small" muted>
            Â© Basestack {new Date().getFullYear()}. All rights reserved.
          </Text>

          <IconButton
            icon={isDarkMode ? "light_mode" : "dark_mode"}
            size="medium"
            onClick={() => {
              toggleDarkMode(!isDarkMode, "bottom-right").then(() => {
                setIsDarkMode(!isDarkMode);
              });
            }}
          />
        </BottomContainer>
      </ContentWrapper>
    </Container>
  );
};

export default Footer;
