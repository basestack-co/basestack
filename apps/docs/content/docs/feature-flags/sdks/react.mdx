---
title: React & NextJS
tags: Basestack, Feature Flags, React, SDK, Hooks, Components, State, Async, Flags, Flag, Initialize, Methods, Reference, Examples
---

import { File, Folder, Files } from 'fumadocs-ui/components/files';


Integration with [React ↗](https://react.dev/?ref=basestack.co) or [NextJS ↗](https://nextjs.org/?ref=basestack.co) helps and facilitates the process of testing and developing features in production and other environments.

## Getting started

For [React ↗](https://react.dev/?ref=basestack.co) and [NextJS ↗](https://nextjs.org/?ref=basestack.co), we chose not to create a separate SDK, as it would
essentially act as a wrapper for our [JavaScript SDK](/feature-flags/sdks/javascript). Instead, we’ll show you
how to fully use the JavaScript SDK within a React or Next.js project.
This approach ensures you can access all the features of a React SDK without
needing to install additional packages. One key advantage is that you retain
complete control over the code, allowing you to extend its functionality with
custom implementations.

#### Examples

- [React (Vite + TypeScript)](https://github.com/basestack-co/examples/tree/main/feature-flags/react-ts)
- [Next,js (Pages Router)](https://github.com/basestack-co/examples/tree/main/feature-flags/nextjs-pages-router)
- [Next,js (App Router)](https://github.com/basestack-co/examples/tree/main/feature-flags/nextjs-app-router)

### Install

First, let's install some packages!

```sh copy npm2yarn
npm install @basestack/flags-js
```

### Environment Variables

Here are the example environment variables needed for this guide. Feel free to rename them according to your project's requirements:

<Callout type="info" emoji="ℹ️">
  When it comes to environment variables, pay attention to the framework you're
  using. For example, `NEXT_PUBLIC_` is specific to Next.js, while in
  [Vite.js](https://vite.dev/?ref=basestack.co), it would be `VITE_`. Example
  `VITE_FEATURE_FLAGS_BASE_URL=` or Node.js `FEATURE_FLAGS_BASE_URL=`
</Callout>

```bash copy title=".env"
# BASESTACK FEATURE FLAGS
NEXT_PUBLIC_FEATURE_FLAGS_BASE_URL="https://flags.basestack.co/api/v1" # or your own URL
NEXT_PUBLIC_FEATURE_FLAGS_PROJECT_KEY=""
NEXT_PUBLIC_FEATURE_FLAGS_ENVIRONMENT_KEY=""
```

### Create a new folder called `feature-flags` in the `src/libs` directory

Inside the `feature-flags` folder, create the following files and folders:

<Files>
  <Folder name="src" defaultOpen>
    <Folder name="libs" defaultOpen>
      <Folder name="feature-flags" defaultOpen>
        <Folder name="hooks" defaultOpen>
          <File name="useFlag.tsx" />
          <File name="useFlags.tsx" />
          <File name="index.ts" />
        </Folder>
        <File name="index.tsx" />
        <File name="server.ts" />
      </Folder>
    </Folder>
  </Folder>
  <File name=".env" />
</Files>

Let's start by creating a Provider. In React, a Provider is a component that allows you to pass down data or state to all components within its tree without having to manually pass props at every level. This will enable you to use the hooks created below anywhere in your components.

In summary, this Provider initializes the `FlagsSDK, collects all the Feature Flags for a given environment to make them available in the context, and helps determine whether the Feature Flags have been initialized.

<Callout type="info" emoji="ℹ️">
  If you are only using React or Next.js in the Pages Router, you can remove the
  `“use client”` directive from the page.
</Callout>

```tsx copy title="src/libs/feature-flags/index.tsx"
"use client";

import React, { createContext, useEffect, useMemo, useState } from "react";
import { FlagsSDK, Flag, SDKConfig } from "@basestack/flags-js";

interface ContextState {
  client: FlagsSDK;
  isInitialized: boolean;
  flags: Flag[];
  error?: Error;
}

interface ProviderProps {
  children: React.ReactNode;
  config: SDKConfig;
  onSuccessfulInit?: (success: boolean) => void;
}

export const createFlagsClient = (config: SDKConfig) => {
  return new FlagsSDK({
    baseURL: config.baseURL,
    projectKey: config.projectKey,
    environmentKey: config.environmentKey,
  });
};

export const FeatureFlagsContext = createContext<ContextState | undefined>(
  undefined,
);

const FeatureFlagsProvider: React.FC<ProviderProps> = ({
  children,
  config,
  onSuccessfulInit,
}) => {
  const [state, setState] = useState<Omit<ContextState, "client">>({
    isInitialized: false,
    flags: [],
  });
  const client = useMemo(() => createFlagsClient(config), [config]);
  const value = useMemo(() => ({ ...state, client }), [state, client]);

  useEffect(() => {
    let isMounted = true;

    const initializeFlags = async () => {
      try {
        const response = await client.getAllFlags();

        if (isMounted) {
          setState((prev) => ({
            ...prev,
            flags: response.flags ?? [],
            isInitialized: true,
          }));
          onSuccessfulInit?.(true);
        }
      } catch (error) {
        if (isMounted) {
          setState((prev) => ({
            ...prev,
            error: error as Error,
            isInitialized: true,
          }));
          onSuccessfulInit?.(false);
        }
      }
    };

    initializeFlags();

    return () => {
      isMounted = false;
    };
  }, [client, onSuccessfulInit]);

  return (
    <FeatureFlagsContext.Provider value={value}>
      {children}
    </FeatureFlagsContext.Provider>
  );
};

export default FeatureFlagsProvider;
```

The `FlagsSDK` is versatile and can be used in various environments, including Node, Serverless environments, API Routes, Server Components, and API Endpoints. To avoid repeatedly initializing the client with environment variables, you can use the file below to simplify the process. Simply call this file as shown in the `usage` example below.

```ts copy title="src/libs/feature-flags/server.ts"
import { FlagsSDK, SDKConfig } from "@basestack/flags-js";

export class ServerFlagsSDK {
  private static instance: FlagsSDK;
  private static config: SDKConfig = {
    baseURL: process.env.NEXT_PUBLIC_FEATURE_FLAGS_BASE_URL,
    projectKey: process.env.NEXT_PUBLIC_FEATURE_FLAGS_PROJECT_KEY!,
    environmentKey: process.env.NEXT_PUBLIC_FEATURE_FLAGS_ENVIRONMENT_KEY!,
    // This is an example of how to use environment variables in Vite.js
    // environmentKey: import.meta.env.VITE_FEATURE_FLAGS_ENVIRONMENT_KEY!,
  };

  public static getInstance(): FlagsSDK {
    if (!ServerFlagsSDK.instance) {
      ServerFlagsSDK.instance = new FlagsSDK(ServerFlagsSDK.config);
    }
    return ServerFlagsSDK.instance;
  }
}
```

This hook, `useFlag`, is used to validate a Feature Flag by searching for a specific flag. It returns the necessary information for the specified flag. See the usage example below. You can customize or modify this hook to fit your requirements.

```tsx copy title="src/libs/feature-flags/hooks/useFlag.tsx"
"use client";

import { useContext, useEffect, useState, useMemo } from "react";
// Feature Flags Context
import { FeatureFlagsContext } from "@/libs/feature-flags";
import { Flag } from "@basestack/flags-js";

export const useFlag = <P = Record<string, unknown>,>(slug: string) => {
  const context = useContext(FeatureFlagsContext);
  const [flag, setFlag] = useState<Flag | null>(null);
  const [error, setError] = useState<Error | null>(null);

  if (!context) {
    throw new Error("useFlag must be used within a FeatureFlagsContext");
  }

  useEffect(() => {
    const fetchFlag = async () => {
      if (!context.isInitialized || !context.client) return;

      try {
        if (context.flags.length > 0) {
          const foundFlag = context.flags.find((f) => f.slug === slug);
          if (foundFlag) {
            setFlag(foundFlag);
          } else {
            setError(new Error(`Flag "${slug}" not found`));
          }
        } else {
          const fetchedFlag = await context.client.getFlag(slug);
          setFlag(fetchedFlag);
        }
      } catch (err) {
        setError(err instanceof Error ? err : new Error(String(err)));
      }
    };

    void fetchFlag();

    return () => {};
  }, [context.isInitialized, context.client, context.flags, slug]);

  return useMemo(
    () => ({
      ...flag,
      error,
      isInitialized: context.isInitialized,
      payload: (flag?.payload as P) ?? null,
    }),
    [error, context.isInitialized, flag],
  );
};
```

This hook, as the name suggests, is used to retrieve all the Feature Flags for a given environment. It can be used to display a list, perform new searches, or apply filtering.

```tsx copy title="src/libs/feature-flags/hooks/useFlags.tsx"
"use client";

import { useContext, useEffect, useState, useMemo } from "react";
// Feature Flags Context
import { FeatureFlagsContext } from "@/libs/feature-flags";
import { Flag } from "@basestack/flags-js";

interface FlagsState {
  flags: Flag[];
}

interface UseFlagsResult extends FlagsState {
  error: Error | null;
  isInitialized: boolean;
}

export const useFlags = (): UseFlagsResult => {
  const context = useContext(FeatureFlagsContext);
  const [{ flags }, setFlags] = useState<FlagsState>({ flags: [] });
  const [error, setError] = useState<Error | null>(null);

  if (!context) {
    throw new Error("useFlags must be used within a FeatureFlagsContext");
  }

  useEffect(() => {
    if (!context.isInitialized || !context.client) return;

    if (context.flags.length > 0) {
      setFlags({ flags: context.flags });
      return;
    }

    context.client
      .getAllFlags()
      .then((result) => setFlags(result))
      .catch((err) =>
        setError(err instanceof Error ? err : new Error(String(err))),
      );

    return () => {};
  }, [
    context.isInitialized,
    context.client,
    context.flags.length,
    context.flags,
  ]);

  return useMemo(
    () => ({
      error,
      isInitialized: context.isInitialized,
      flags,
    }),
    [flags, context.isInitialized, error],
  );
};
```

Finally, export all the created hooks to make them easy to import. For example: `import { useFlag, useFlags } from "@/libs/feature-flags/hooks";`

```ts copy title="src/libs/feature-flags/hooks/index.ts"
export * from "./useFlag";
export * from "./useFlags";
```

### Add the `FeatureFlagsProvider` to your App

The example below demonstrates how to add the `FeatureFlagsProvider` to your Next.js app using the App Router. It can also be used in the `_app.tsx` file or a React SPA app.

```tsx copy title="src/app/layout.tsx"
import React from "react";
import "./globals.css";
// Feature Flags
import FeatureFlagsProvider from "@/libs/feature-flags";

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body>
        <FeatureFlagsProvider
          config={{
            baseURL: process.env.NEXT_PUBLIC_FEATURE_FLAGS_BASE_URL,
            projectKey: process.env.NEXT_PUBLIC_FEATURE_FLAGS_PROJECT_KEY!,
            environmentKey:
              process.env.NEXT_PUBLIC_FEATURE_FLAGS_ENVIRONMENT_KEY!,
          }}
        >
          {children}
        </FeatureFlagsProvider>
      </body>
    </html>
  );
}
```

Your app is now fully equipped to take advantage of feature flags and other powerful functionalities offered by Basestack. To maximize the potential of Basestack, refer to the instructions provided in the supported hooks and components documentation.

### Usage

The examples demonstrate how to integrate the provided code into your project. If you find that an important example is missing, please open an issue at the following [link](https://github.com/basestack-co/basestack/issues).

#### React Component

Use hooks in React components to fetch a single Feature Flag or all Feature Flags.

To support types in the payload defined in the Dashboard within the _useFlag hook_, use the following syntax: `useFlag<YourPayloadType>`.

```tsx copy title="src/app/page.tsx"
"use client";

// Feature Flags Hooks
import { useFlag, useFlags } from "@/libs/feature-flags/hooks";

export default function Home() {
  const count = useFlag<{ text: string }>("count");
  const data = useFlags();

  return (
    <div>
      <main>
        {count.enabled && (
          <div>
            <a>
              {!count.enabled
                ? "Count is not enabled"
                : `${count?.payload?.text} 0`}
            </a>
          </div>
        )}

        <h3>All the available flags</h3>
        <ul>
          {data.flags.map((flag, index) => {
            return (
              <li
                key={index}
              >{`Slug: ${flag.slug} Enabled: ${flag.enabled}`}</li>
            );
          })}
        </ul>
      </main>
    </div>
  );
}
```

#### API Route Handler (App Router)

On the server side, use the _server file_ that already has the `FlagsSDK` initialized. Use this file wherever the server is needed, such as in API Routes, Server Components, API Endpoints, etc.

```ts copy title="src/app/api/count/route.ts"
import { ServerFlagsSDK } from "@/libs/feature-flags/server";

export async function GET() {
  const flagsClient = ServerFlagsSDK.getInstance();

  const flag = await flagsClient.getFlag("count");
  const data = await flagsClient.getAllFlags();

  return Response.json({ data, flag });
}
```

#### API Endpoint (Pages Router)

Here's an example of using the server instance in a Next.js Pages Router API endpoint:

```ts copy title="src/pages/api/hello.ts"
import type { NextApiRequest, NextApiResponse } from "next";
import { ServerFlagsSDK } from "@/libs/feature-flags/server";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse,
) {
  const flagsClient = ServerFlagsSDK.getInstance();

  const flag = await flagsClient.getFlag("count");
  const data = await flagsClient.getAllFlags();

  res.status(200).json({ data, flag });
}
```

#### Server Component

You can use the server instance in any server-side context, including `Server Components` in Next.js.

```tsx copy title="src/app/count/page.tsx"
import { ServerFlagsSDK } from "@/libs/feature-flags/server";

export default async function Page() {
  const flagsClient = ServerFlagsSDK.getInstance();
  const data = await flagsClient.getAllFlags();

  return (
    <div>
      <h3>All the available flags in server component</h3>
      <ul>
        {data.flags.map((flag, index) => {
          return (
            <li key={index}>{`Slug: ${flag.slug} Enabled: ${flag.enabled}`}</li>
          );
        })}
      </ul>
    </div>
  );
}
```

#### Direct Call with Context

You can access the client directly through the context, providing more control when working with useEffect, onClick handlers, or other scenarios where direct access is needed.

```tsx copy title="src/app/posts/page.tsx"
"use client";

import { useState, useEffect, useContext } from "react";
// Feature Flags Context
import { FeatureFlagsContext } from "@/libs/feature-flags";
import { Flag } from "@basestack/flags-js";

export default function Posts() {
  const [flags, setFlags] = useState<Flag[]>([]);
  const context = useContext(FeatureFlagsContext);

  useEffect(() => {
    async function fetchAllFlags() {
      const data = await context?.client.getAllFlags();
      setFlags(data?.flags ?? []);
    }
    fetchAllFlags();
  }, []);

  if (!flags) return <div>Loading...</div>;

  return (
    <ul>
      {flags.map((flag, index) => {
        return (
          <li key={index}>{`Slug: ${flag.slug} Enabled: ${flag.enabled}`}</li>
        );
      })}
    </ul>
  );
}
```

## Initialisation options

For details regarding the initialization options of @basestack/flags-js, please refer to the documentation related to the [Javascript SDK](/feature-flags/sdks/javascript#initialisation-options).

## Troubleshooting

### Why are my components rendering twice?

It's normal. Check if your App/Component is wrapped inside [React.StrictMode ↗](https://react.dev/reference/react/StrictMode?ref=basestack.co). The purpose of `React.StrictMode` is to help identify potential problems in your application by highlighting some unsafe practices or potential bugs during development.

When you wrap a component in `React.StrictMode`, it can cause some components to be rendered twice. This is expected behavior and is not specific to your code. Rendering twice is intentional and done by React on purpose.

```js copy title="main.ts"
const root = ReactDOM.createRoot(document.getElementById("root"));

root.render(
  <React.StrictMode>
    <FeatureFlagsProvider
      config={{
        baseURL: process.env.NEXT_PUBLIC_FEATURE_FLAGS_BASE_URL,
        projectKey: process.env.NEXT_PUBLIC_FEATURE_FLAGS_PROJECT_KEY!,
        environmentKey:
          process.env.NEXT_PUBLIC_FEATURE_FLAGS_ENVIRONMENT_KEY!,
      }}
    >
        <App />
    </FeatureFlagsProvider>
  </React.StrictMode>,
);
```
