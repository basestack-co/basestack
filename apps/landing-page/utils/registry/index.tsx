"use client";

import Footer from "components/Footer";
import GlobalNavigation from "components/GlobalNavigation";
import { usePathname } from "next/navigation";
import React, { Fragment, useMemo } from "react";
// Components
import { Toaster } from "sonner";
// Registries
import StyledComponentsRegistry from "./StyledComponentsRegistry";
// Fonts
import "material-symbols/rounded.css";

const Registry = ({ children }: { children: React.ReactNode }) => {
  const pathname = usePathname();
  const isProductPage = useMemo(
    () => pathname.includes("/product"),
    [pathname],
  );

  const hasGlobalNavigation = useMemo(
    () =>
      !["/legal/privacy", "/legal/cookies", "/legal/terms"].includes(pathname),
    [pathname],
  );

  return (
    <StyledComponentsRegistry>
      <Fragment>
        {hasGlobalNavigation && <GlobalNavigation isSticky={!isProductPage} />}
        {children}
        <Footer />
      </Fragment>
      <Toaster visibleToasts={9} closeButton={false} />
    </StyledComponentsRegistry>
  );
};

export default Registry;
