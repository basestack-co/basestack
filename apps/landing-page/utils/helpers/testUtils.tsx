import theme from "@basestack/design-system/theme/lightTheme";
// Styles
import isPropValid from "@emotion/is-prop-valid";
import { RenderOptions, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

interface WithChildrenProps {
  children: React.ReactNode;
}

interface AllProvidersProps extends WithChildrenProps {
  initialState?: {};
}

const AllProviders = ({ children, initialState = {} }: AllProvidersProps) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const renderWithAllProviders = (
  ui: React.ReactElement,
  initialState?: {},
  options?: Omit<RenderOptions, "wrapper">,
) =>
  render(ui, {
    // @ts-ignore
    wrapper: (props) => <AllProviders initialState={initialState} {...props} />,
    ...options,
  });

const WithThemeProvider = ({ children }: WithChildrenProps) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
  // @ts-ignore
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithAllProviders, renderWithTheme };
