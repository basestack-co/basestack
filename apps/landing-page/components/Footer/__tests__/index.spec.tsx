import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import Footer from "..";

describe("Footer tests", () => {
  afterEach(cleanup);

  test("render Footer correctly", () => {
    const { asFragment } = renderWithTheme(<Footer />);

    expect(asFragment()).toMatchSnapshot();
  });
});
