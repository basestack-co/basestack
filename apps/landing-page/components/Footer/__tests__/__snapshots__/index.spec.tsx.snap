// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer tests should render Footer correctly 1`] = `
<DocumentFragment>
  .c6 {
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 500;
  font-family: roboto;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  word-break: break-word;
  color: #CBCBCB;
}

.c10 {
  font-size: 1rem;
  line-height: 1.5rem;
  font-weight: 500;
  font-family: roboto;
  font-weight: 400;
  margin-bottom: 0.5rem;
  font-family: 'Roboto',sans-serif;
  word-break: break-word;
  color: #CBCBCB;
}

.c3 {
  font-size: 1.25rem;
  line-height: 1.875rem;
  font-weight: 500;
  font-family: roboto;
  margin-bottom: 0.5rem;
  font-family: 'Roboto',sans-serif;
  word-break: break-word;
  color: #CBCBCB;
}

.c13 {
  height: 2.625rem;
  padding: 0 1rem;
  font-size: 1rem;
  position: relative;
  border: none;
  display: flex;
  align-items: center;
  font-weight: 500;
  cursor: pointer;
  border-radius: 4px;
  text-decoration: none;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out,border-color 0.1s ease-in-out;
  background-color: #276EF1;
  color: #FFFFFF;
  width: medium;
  height: medium;
  width: fit-content;
}

.c13:hover:not(:active, :disabled) {
  background-color: #0e55d7;
}

.c13 .material-symbols-sharp,
.c13 .material-symbols-rounded {
  color: inherit;
}

.c13 svg,
.c13 svg g {
  fill: #FFFFFF;
}

.c13:disabled {
  cursor: default;
  opacity: 0.6;
}

.c0 {
  display: flex;
  padding: 3.125rem 1.25rem;
  background-color: #1F1F1F;
}

.c1 {
  display: flex;
  justify-content: space-between;
  max-width: 90rem;
  width: 100%;
  margin: 0 auto;
}

.c2 {
  display: flex;
  flex-direction: column;
}

.c4 {
  display: flex;
  align-items: center;
}

.c5 {
  display: inline-flex;
  height: 2.25rem;
  align-items: center;
}

.c7 {
  display: flex;
  margin-left: 1rem;
}

.c8 {
  display: inline-flex;
  height: 2.25rem;
  align-items: center;
  padding: 0 1rem;
}

.c8:hover {
  cursor: pointer;
}

.c9 {
  display: flex;
  flex-direction: column;
  max-width: 20rem;
  width: 100%;
}

.c11 {
  display: flex;
}

.c12 {
  font-size: 1rem;
  height: 2.625rem;
  width: 100%;
  border-radius: 0.25rem;
  border: none;
  margin-right: 1rem;
  padding: 0 1rem;
  color: #F6F6F6;
  background-color: #333333;
}

.c12::placeholder {
  color: #AFAFAF;
}

.c12:focus {
  outline: 2px solid #757575;
}

@media screen and (max-width: 575.98px) {
  .c1 {
    flex-direction: column;
    justify-content: initial;
  }
}

@media screen and (max-width: 575.98px) {
  .c2 {
    margin-top: 1.875rem;
    order: 1;
  }
}

@media screen and (max-width: 991.98px) {
  .c4 {
    flex-direction: column;
    align-items: initial;
    justify-content: center;
  }
}

@media screen and (max-width: 991.98px) {
  .c5 {
    margin: 0.5rem 0;
  }
}

@media screen and (max-width: 991.98px) {
  .c7 {
    flex-direction: column;
    margin-left: 0;
  }
}

@media screen and (max-width: 991.98px) {
  .c8 {
    padding: 0;
    margin: 0 0 0.5rem 0;
  }
}

<footer
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <h2
          class="c3"
          color="#CBCBCB"
          font-family="roboto"
        >
          The Open-Source Stack for Developers and Startups
        </h2>
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <p
              class="c6"
              color="#CBCBCB"
              font-family="roboto"
              font-weight="400"
            >
              Â© Basestack 2024. All rights reserved.
            </p>
          </div>
          <ul
            class="c7"
          >
            <li
              class="c8"
            >
              <p
                class="c6"
                color="#CBCBCB"
                font-family="roboto"
                font-weight="400"
              >
                Privacy Policy
              </p>
            </li>
            <li
              class="c8"
            >
              <p
                class="c6"
                color="#CBCBCB"
                font-family="roboto"
                font-weight="400"
              >
                Cookies Policy
              </p>
            </li>
            <li
              class="c8"
            >
              <p
                class="c6"
                color="#CBCBCB"
                font-family="roboto"
                font-weight="400"
              >
                Term of Use
              </p>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="c9"
      >
        <p
          class="c10"
          color="#CBCBCB"
          font-family="roboto"
          font-weight="400"
        >
          Get updates directly to your Inbox
        </p>
        <div
          class="c11"
        >
          <input
            class="c12"
            name="email"
            placeholder="Enter your email"
            type="email"
            value=""
          />
          <button
            class="c13"
            data-testid="button"
          >
            <span
              class=""
            >
              Subscribe
            </span>
          </button>
        </div>
      </div>
    </div>
  </footer>
</DocumentFragment>
`;
